great_expectations_cloud/agent/actions/draft_datasource_config_action.py

```diff
@@ -5,19 +5,34 @@
 
 from great_expectations.compatibility import pydantic
 from great_expectations.core.http import create_session
+from great_expectations.datasource.fluent.interfaces import TestConnectionError
 from typing_extensions import override
 
 from great_expectations_cloud.agent.actions import ActionResult, AgentAction
 from great_expectations_cloud.agent.config import (
     GxAgentEnvVars,
     generate_config_validation_error_text,
 )
+from great_expectations_cloud.agent.exceptions import raise_with_error_code
 from great_expectations_cloud.agent.models import DraftDatasourceConfigEvent
 
 
 class DraftDatasourceConfigAction(AgentAction[DraftDatasourceConfigEvent]):
     @override
     def run(self, event: DraftDatasourceConfigEvent, id: str) -> ActionResult:
+        try:
+            return self.check_draft_datasource_config(event=event, id=id)
+        except TestConnectionError as e:
+            if "Incorrect username or password was specified" in str(
```

Comment:
string matching is brittle, is there an alternate way to discern between exceptions here?

Reply:
Yes, I would suggest inspecting the `e.__cause__`

Reply:
Absolutely - very soon in GX Core 1.0 we will add more error specificity. That way we can catch specific errors instead of catching and string matching to discern. But if there is a better way in the meantime without changing GX Core 0.18.x, I'm open to suggestions!

Reply:
I'm not finding anything better than a string error msg in `e.__cause__`, is there something I'm missing?:

```
-> status = build_failed_job_completed_status(error)
(Pdb) error.__cause__
TestConnectionError('Attempt to connect to datasource failed with the following error message: (snowflake.connector.errors.DatabaseError) 250001 (08001): None: Failed to connect to DB: <DB URL>. Incorrect username or password was specified.\n(Background on this error at: https://sqlalche.me/e/14/4xp6)')
(Pdb) dir(error.__cause__)
['__cause__', '__class__', '__context__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__setstate__', '__sizeof__', '__str__', '__subclasshook__', '__suppress_context__', '__traceback__', '__weakref__', 'args', 'with_traceback']
(Pdb) error
GXCoreError('Attempt to connect to datasource failed with the following error message: (snowflake.connector.errors.DatabaseError) 250001 (08001): None: Failed to connect to DB: <DB URL>. Incorrect username or password was specified.\n(Background on this error at: https://sqlalche.me/e/14/4xp6)')
(Pdb) error.__cause__.__cause__
DatabaseError('(snowflake.connector.errors.DatabaseError) 250001 (08001): None: Failed to connect to DB: <DB URL>. Incorrect username or password was specified.')
(Pdb) dir(error.__cause__.__cause__)
['__cause__', '__class__', '__context__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__setstate__', '__sizeof__', '__str__', '__subclasshook__', '__suppress_context__', '__traceback__', '__unicode__', '__weakref__', '_code_str', '_message', '_sql_message', 'add_detail', 'args', 'code', 'connection_invalidated', 'detail', 'hide_parameters', 'instance', 'ismulti', 'orig', 'params', 'statement', 'with_traceback']
(Pdb) error.__cause__.__cause__.code
'4xp6'
(Pdb) error.__cause__.__cause__.detail
[]
(Pdb) error.__cause__.__cause__.params
(Pdb) error.__cause__.__cause__.statement
(Pdb) error.__cause__.__cause__._code_str
<bound method HasDescriptionCode._code_str of DatabaseError('(snowflake.connector.errors.DatabaseError) 250001 (08001): None: Failed to connect to DB: <DB URL>. Incorrect username or password was specified.')>
(Pdb) error.__cause__.__cause__._code_str()
'(Background on this error at: https://sqlalche.me/e/14/4xp6)'
(Pdb) error.__cause__.__cause__.instance
<bound method DBAPIError.instance of <class 'sqlalchemy.exc.DatabaseError'>>
(Pdb) error.__cause__.__cause__.instance()
*** TypeError: instance() missing 4 required positional arguments: 'statement', 'params', 'orig', and 'dbapi_base_err'
(Pdb) error.__cause__.__cause__.orig
250001 (08001): None: Failed to connect to DB: <DB URL>. Incorrect username or password was specified.
(Pdb) error.__cause__.__cause__.params

```

Reply:
Looks like snowflake errors contain `errno`

```
ex.orig.errno == 250001
```

https://github.com/search?q=repo%3Asnowflakedb%2Fsnowflake-sqlalchemy%20250001&type=code

Reply:
@anthonyburdi you can do `isinstance` checks on the `e.__cause__` rather than relying on string matching.

Reply:
It is a very valid concern, I'll add a comment in the string matching part itself describing why we are doing it this way here and that we shouldn't follow this pattern going forward.

Reply:
See https://github.com/great-expectations/cloud/pull/160/commits/4f66d90fd149aa61140025480538e4a8c0762dc7